File: 2020/15/Rust/src/main.rs
use std::env;
use std::fs;

const ITERATIONS_1: usize = 2020;
const ITERATIONS_2: usize = 30_000_000;

type Game = Vec<(usize, usize)>;

fn parse_line(line: &str) -> Vec<usize> {
    line.split(',')
        .map(|x| {
            if let Some(y) = x.strip_suffix('\n') {
                y.parse().unwrap()
            } else {
                x.parse().unwrap()
            }
        })
        .collect()
}

fn parse_input(input: &str) -> Vec<Vec<usize>> {
    input
        .lines()
        .filter(|x| !x.is_empty())
        .map(|y| parse_line(y))
        .collect()
}

fn init_game(numbers: &[usize], mut game: Game) -> Game {
    for (index, number) in numbers.iter().enumerate() {
        let index = index + 1;
        game[*number] = (index, index);
    }
    game
}

fn step_game(
    offset: usize,
    end: usize,
    mut current: usize,
    mut game: Game,
) -> usize {
    let mut before_last;
    let mut last;
    let mut delta;

    for round in offset..=end {
        before_last = game[current].0;
        last = game[current].1;

        delta = last - before_last;

        before_last = game[delta].0;
        last = game[delta].1;

        before_last = {
            if before_last == 0 {
                round
            } else {
                last
            }
        };
        last = round;

        game[delta] = (before_last, last);

        current = delta;
    }

    current
}

fn solve(numbers: &[usize], iterations: &usize) -> usize {
    let game = vec![(0, 0); *iterations];
    let offset = numbers.len() + 1;
    let first = numbers.last().unwrap();

    let game = init_game(&numbers, game);

    step_game(offset, *iterations, *first, game)
}

fn show(message: &str, iterations: usize, input: &[Vec<usize>]) {
    let solver = |iterations| {
        input
            .iter()
            .map(|x| solve(x, &iterations))
            .collect::<Vec<_>>()
    };

    let printer = |vector: Vec<usize>| {
        if vector.len() == 1 {
            println!("  Part {}: {}", message, vector[0]);
        } else {
            println!("  Part {}: {:?}", message, vector);
        }
    };

    let solution = solver(iterations);
    printer(solution);
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    println!("File: {}", path);
    show("1", ITERATIONS_1, &parsed);
    show("2", ITERATIONS_2, &parsed);
}

fn main() {
    env::args()
        .skip(1)
        .collect::<Vec<_>>()
        .iter()
        .for_each(|x| output(x));
}

File: 2020/15/Haskell/src/Main.hs
import qualified Data.Text                     as T
import           Data.Vector.Unboxed.Mutable    ( IOVector )
import qualified Data.Vector.Unboxed.Mutable   as MV
import           System.Environment
import           Text.Printf

type Game = IOVector (Int, Int)

parseInput :: String -> [Int]
parseInput = map (read . T.unpack . T.strip) . T.splitOn (T.pack ",") . T.pack

initGame :: [Int] -> Game -> IO Game
initGame n v = mapM_ write base >> return v
 where
  base  = zip n [ (t, t) | t <- [1 ..] ]
  write = uncurry (MV.write v)

stepGame :: Int -> Int -> Int -> Game -> IO Int
stepGame i0 i n v
  | i == i0 = return n
  | otherwise = do
    (f, l) <- MV.read v n
    let n' = l - f
    (fn, ln) <- MV.read v n'
    let (fn', ln') | fn == 0   = (i, i)
                   | otherwise = (ln, i)
    MV.write v n' (fn', ln')
    stepGame i0 (i + 1) n' v

solve :: Int -> [Int] -> IO Int
solve i n = MV.replicate i (0, 0) >>= initGame n >>= stepGame (i + 1) offset n0
 where
  n0     = last n
  offset = length n + 1

output :: String -> IO ()
output path = do
  input <- map parseInput . filter (not . null) . lines <$> readFile path
  printf "File: %s\n" path
  part1 <- mapM (solve 2020) input
  printf "  Part 1: %s\n" . extract $ part1
  part2 <- mapM (solve 30000000) input
  printf "  Part 2: %s\n" . extract $ part2
 where
  singleton [_] = True
  singleton _   = False
  extract x | singleton x = show . head $ x
            | otherwise   = show x

main :: IO ()
main = getArgs >>= mapM_ output
